crates/gpui/src/platform/linux/wayland/client.rs

```diff
@@ -422,32 +442,57 @@ impl Dispatch<wl_keyboard::WlKeyboard, ()> for WaylandClientState {
                     (key_utf32 == 0 || (key_utf32 >= 32 && key_utf32 != 127)).then_some(key_utf8);
 
                 let focused_window = &state.keyboard_focused_window;
-                if let Some(focused_window) = focused_window {
-                    match key_state {
-                        wl_keyboard::KeyState::Pressed => {
-                            focused_window.handle_input(PlatformInput::KeyDown(KeyDownEvent {
-                                keystroke: Keystroke {
-                                    modifiers: state.modifiers,
-                                    key,
-                                    ime_key,
-                                },
-                                is_held: false, // todo!(linux)
-                            }));
+                let Some(focused_window) = focused_window else {
+                    return;
+                };
+                match key_state {
+                    wl_keyboard::KeyState::Pressed => {
+                        let input = PlatformInput::KeyDown(KeyDownEvent {
+                            keystroke: Keystroke {
+                                modifiers: state.modifiers,
+                                key,
+                                ime_key,
+                            },
+                            is_held: false, // todo!(linux)
+                        });
+
+                        focused_window.handle_input(input);
+
+                        if !key_sym.is_modifier_key() {
+                            state.repeat.current_id += 1;
+                            state.repeat.current_keysym = Some(key_sym);
+
+                            let rate = state.repeat.rate;
+                            let delay = state.repeat.delay;
+                            let id = state.repeat.current_id;
+                            let keysym = state.repeat.current_keysym;
+
+                            // state.platform_inner.foreground_executor.spawn(async move {
+                            //     Timer::after(Duration::from_millis(delay as u64)).await;
+                            //     if id == state.repeat.current_id &&
+                            //          state.repeat.current_keysym.is_some() {
+                            //
+                            //          focused_window.handle_input(input);
+                            //     }
+                            // });
```

Comment:
I'm trying to do this, but I haven't been able to figure out how to make the lifetimes work yet. Suggestions welcome.

Reply:
Try creating the timer above the spawn call, then moving it in and awaiting it!

Reply:
You'll also need to get some kind of Arc<Mutex<State>> to solve your lifetime issues.